src/wpt/harness.ts

```diff
@@ -219,15 +427,72 @@ globalThis.subsetTestByKey = (
   return testType(testCallback, testMessage);
 };
 
-globalThis.promise_test = (func, name, _properties): void => {
+globalThis.promise_test = (func, name, properties): void => {
   if (!shouldRunTest(name)) {
     return;
   }
 
-  try {
-    globalThis.promises[name] = func.call(this);
-  } catch (err) {
-    globalThis.errors.push(new AggregateError([err], name));
+  const testCase = new Test(name, properties);
+  const promise = testCase.step(func, testCase, testCase);
+
+  if (testCase.error) {
+    globalThis.errors.push(testCase.error);
+  }
+
+  if (promise instanceof Promise) {
+    globalThis.promises.push(
+      promise
+        .catch((err: unknown) => {
+          globalThis.errors.push(new AggregateError([err], name));
+        })
+        .then(() => {
+          if (testCase.error) {
+            globalThis.errors.push(testCase.error);
+          }
```

Comment:
line 438 and 449 does the same exact thing. doesn't this duplicate the errors?

Reply:
Yeah this is a bit messy. I'll try to rework this.